ios: it seems we cannot use `thread_local` storage

I was cross compiling MK `fix/dash` branch, with #1314 applied, and discovered that `thread_local` is not supported by some target:

```
configure: WARNING: using cross tools not prefixed with host triplet
configure: WARNING: skipped the ca-bundle detection when cross-compiling
src/libmeasurement_kit/ooni/utils.cpp:22:12: error: thread-local storage is not supported for the current target
    static thread_local Var<GeoipCache> singleton(new GeoipCache);
           ^
1 error generated.
make[1]: *** [src/libmeasurement_kit/ooni/utils.lo] Error 1
make: *** [all] Error 2
```

From the log, it was not obvious which was the target. Yet, we probably need to replace the solution based on thread local storage with another solution based instead on locking.

This is interesting because it probably means that we don't need to concern ourselves too much with solving clearly-thread-local-storage issues in #1313.

(Even if thread local storage was not supported for an "irrelevant" arch like `i386`, I'd say it's better to avoid using it &mdash; also keeping into account that I want a `v0.7.1` release by end of next week.)

--- a/build/ios/cross
+++ b/build/ios/cross
@@ -12,7 +12,7 @@ shift
 ARCH=$1
 shift
 
-MINIOSVERSION="8.0"
+MINIOSVERSION="9.0"
 if [ "$PLATFORM" = "iphoneos" ]; then
     EXTRA_CONFIG="--host=arm-apple-darwin --target=arm-apple-darwin --disable-shared"
     MINVERSION="-miphoneos-version-min=$MINIOSVERSION"

--- a/include/private/ooni/utils.hpp
+++ b/include/private/ooni/utils.hpp
@@ -42,7 +42,7 @@ class GeoipDatabase {
 
 class GeoipCache {
   public:
-    static Var<GeoipCache> global();
+    static Var<GeoipCache> thread_local_instance();
 
     Var<GeoipDatabase> get(std::string path, bool &did_open);
 

--- a/measurement_kit.podspec
+++ b/measurement_kit.podspec
@@ -19,6 +19,6 @@ Pod::Spec.new do |s|
   s.prepare_command = <<-CMD
     ./build/ios/library
   CMD
-  s.platform = :ios, "7.1"
+  s.platform = :ios, "9.0"
   s.vendored_framework = "build/ios/Frameworks/*.framework"
 end

--- a/src/libmeasurement_kit/nettests/runnable.cpp
+++ b/src/libmeasurement_kit/nettests/runnable.cpp
@@ -205,7 +205,7 @@ void Runnable::geoip_lookup(Callback<> cb) {
                                             std::string{});
             if (save_cc and country_path != "") {
                 try {
-                    probe_cc = *GeoipCache::global()
+                    probe_cc = *GeoipCache::thread_local_instance()
                        ->resolve_country_code(country_path, ip, logger);
                 } catch (const Error &err) {
                     logger->warn("cannot lookup country code: %s",
@@ -221,7 +221,7 @@ void Runnable::geoip_lookup(Callback<> cb) {
             auto asn_path = options.get("geoip_asn_path", std::string{});
             if (save_asn and asn_path != "") {
                 try {
-                    probe_asn = *GeoipCache::global()
+                    probe_asn = *GeoipCache::thread_local_instance()
                         ->resolve_asn(asn_path, ip, logger);
                 } catch (const Error &err) {
                     logger->warn("cannot lookup asn: %s",

--- a/src/libmeasurement_kit/ooni/utils.cpp
+++ b/src/libmeasurement_kit/ooni/utils.cpp
@@ -18,8 +18,8 @@ void resolver_lookup(Callback<Error, std::string> callback, Settings settings,
     resolver_lookup_impl(callback, settings, reactor, logger);
 }
 
-/* static */ Var<GeoipCache> GeoipCache::global() {
-    static Var<GeoipCache> singleton(new GeoipCache);
+/* static */ Var<GeoipCache> GeoipCache::thread_local_instance() {
+    static thread_local Var<GeoipCache> singleton(new GeoipCache);
     return singleton;
 }
 

--- a/src/libmeasurement_kit/ooni/web_connectivity.cpp
+++ b/src/libmeasurement_kit/ooni/web_connectivity.cpp
@@ -190,19 +190,15 @@ static void compare_dns_queries(Var<Entry> entry,
     std::set<std::string> ctrl_asns;
 
     std::string asn_p = options.get("geoip_asn_path", std::string{});
-    logger->debug("Creating..");
-    // TODO: perhaps here we could use GeoipCache::global()
-    GeoipDatabase ip_location(asn_p);
+    auto ip_location = GeoipCache::thread_local_instance()->get(asn_p);
     for (auto exp_addr : exp_addresses) {
-        logger->debug("expaddr");
-        ErrorOr<std::string> asn = ip_location.resolve_asn(exp_addr);
+        ErrorOr<std::string> asn = ip_location->resolve_asn(exp_addr);
         if (asn && asn.as_value() != "AS0") {
             exp_asns.insert(asn.as_value());
         }
     }
-    logger->debug("control");
     for (auto ctrl_addr : ctrl_addresses) {
-        ErrorOr<std::string> asn = ip_location.resolve_asn(ctrl_addr);
+        ErrorOr<std::string> asn = ip_location->resolve_asn(ctrl_addr);
         if (asn && asn.as_value() != "AS0") {
             ctrl_asns.insert(asn.as_value());
         }

--- a/test/ooni/utils.cpp
+++ b/test/ooni/utils.cpp
@@ -137,19 +137,19 @@ TEST_CASE("ip lookup works") {
 }
 
 TEST_CASE("geoip works") {
-    auto asn = ooni::GeoipCache::global()->resolve_asn(
+    auto asn = ooni::GeoipCache::thread_local_instance()->resolve_asn(
             "GeoIPASNum.dat",
             "130.192.16.172"
     );
-    auto cname = ooni::GeoipCache::global()->resolve_country_name(
+    auto cname = ooni::GeoipCache::thread_local_instance()->resolve_country_name(
             "GeoIP.dat",
             "130.192.16.172"
     );
-    auto cc = ooni::GeoipCache::global()->resolve_country_code(
+    auto cc = ooni::GeoipCache::thread_local_instance()->resolve_country_code(
             "GeoIP.dat",
             "130.192.16.172"
     );
-    auto city = ooni::GeoipCache::global()->resolve_city_name(
+    auto city = ooni::GeoipCache::thread_local_instance()->resolve_city_name(
             "GeoLiteCity.dat",
             "130.192.16.172"
     );
@@ -160,64 +160,64 @@ TEST_CASE("geoip works") {
 }
 
 TEST_CASE("geoip memoization works") {
-    ooni::GeoipCache::global()->invalidate(); // Start clean
+    ooni::GeoipCache::thread_local_instance()->invalidate(); // Start clean
 
     // Open more then once. After the first open we should not really open.
-    auto gi = ooni::GeoipCache::global()->get(
+    auto gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoIP.dat");
     bool first_open;
 
     first_open = true;
-    gi = ooni::GeoipCache::global()->get(
+    gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoIP.dat", first_open);
     REQUIRE(first_open == false);
 
     // Repeat two more times to make sure behavior is consistent
 
     first_open = true;
-    gi = ooni::GeoipCache::global()->get(
+    gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoIP.dat", first_open);
     REQUIRE(first_open == false);
 
     first_open = true;
-    gi = ooni::GeoipCache::global()->get(
+    gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoIP.dat", first_open);
     REQUIRE(first_open == false);
 
     // Make sure that, if we change at least one file name, we reopen all
 
     first_open = false;
-    gi = ooni::GeoipCache::global()->get(
+    gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoLiteCity.dat", first_open);
     REQUIRE(first_open == true);
 
     // Make sure that, if we close, then of course we reopen
 
-    ooni::GeoipCache::global()->invalidate();
+    ooni::GeoipCache::thread_local_instance()->invalidate();
 
     first_open = false;
-    gi = ooni::GeoipCache::global()->get(
+    gi = ooni::GeoipCache::thread_local_instance()->get(
         "GeoLiteCity.dat", first_open);
     REQUIRE(first_open == true);
 
 }
 
 TEST_CASE("IpLocation::resolve_countr_code() deals with nonexistent database") {
-    REQUIRE((ooni::GeoipCache::global()->resolve_country_code(
+    REQUIRE((ooni::GeoipCache::thread_local_instance()->resolve_country_code(
                     "invalid.dat", "8.8.8.8"
                 ).as_error()
              == ooni::GeoipDatabaseOpenError()));
 }
 
 TEST_CASE("IpLocation::resolve_countr_name() deals with nonexistent database") {
-    REQUIRE((ooni::GeoipCache::global()->resolve_country_name(
+    REQUIRE((ooni::GeoipCache::thread_local_instance()->resolve_country_name(
                     "invalid.dat", "8.8.8.8"
                 ).as_error()
              == ooni::GeoipDatabaseOpenError()));
 }
 
 TEST_CASE("IpLocation::resolve_asn() deals with nonexistent database") {
-    REQUIRE((ooni::GeoipCache::global()->resolve_asn(
+    REQUIRE((ooni::GeoipCache::thread_local_instance()->resolve_asn(
                     "invalid.dat", "8.8.8.8"
                 ).as_error()
              == ooni::GeoipDatabaseOpenError()));

